Change Namespace for Stored Script to Only Use Id

jdconrad opened the pull request:
Currently, stored scripts use a namespace of (lang, id) to be put, get, deleted, and executed.  This is not necessary since the lang is stored with the stored script.  A user should only have to specify an id to use a stored script.  This change makes that possible while keeping backwards compatibility with the previous namespace of (lang, id).  Anywhere the previous namespace is used will log deprecation warnings.  The goal is to backport this change to 5.2 and then remove the previous namespace altogether in 6.0 giving users the opportunity to change their scripts to the new namespace.

Why is this change so large?

This change seems huge, but it's not really.  The main files changed are ScriptMetaData and ScriptService with StoredScriptSource being added.  Secondarily, most of the Request files have changed a bit to support the new/old namespaces.  Many lines that have changed are added JavaDocs and some new tests.  Also maintaining back compat eats up some lines for XContent and the wire format.

What is the new behavior?

When a user specifies a stored script, that script will be stored under both the new namespace and old namespace.

Take for example script 'A' with lang 'L0' and data 'D0'.  If we add script 'A' to the empty set, the scripts map will be ["A" -- D0, "A#L0" -- D0].  If a script 'A' with lang 'L1' and data 'D1' is then added, the scripts map will be ["A" -- D1, "A#L1" -- D1, "A#L0" -- D0].

When a user deletes a stored script, that script will be deleted from both the new namespace (if it exists) and the old namespace.

Take for example a scripts map with {"A" -- D1, "A#L1" -- D1, "A#L0" -- D0}. If a script is removed specified by an id 'A' and lang null then the scripts map will be {"A#L0" -- D0}.  To remove the final script, the deprecated namespace must be used, so an id 'A' and lang 'L0' would need to be specified.

When a user gets/executes a stored script, if the new namespace is used then the script will be retrieved/executed using only 'id', and if the old namespace is used then the script will be retrieved/executed using 'id' and 'lang'

clintongormley commented:
@jdconrad Have you added deprecation logging for when the old form (including lang) has been used?

jdconrad commented:
@clintongormley Sorry about missing the deprecation tag!  I think I always miss one.  To answer your question I added deprecation logging everywhere user-input can come in from that uses lang.

martijnvg commented on modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml:
```diff
@@ -1,14 +1,20 @@
 ---
 "Indexed script":
```
not part of this change, but this should be renamed to `"Stored script"` :)

jdconrad replied:
Done.

martijnvg commented on modules/lang-painless/src/test/resources/rest-api-spec/test/painless/16_update2.yaml:
```diff
@@ -28,6 +34,8 @@
 
   - do:
+     warnings:
+       - 'specifying lang [painless] as part of the url path is deprecated'
```
Also I think we should update the yaml specs to just describe the id and update the yaml tests accordingly.
So that the new behavior gets tested too.

We can add `get_script_deprecated.json`, `delete_script_deprecated.json` etc. to test that the old behaviour and that warnings get returned in the response.

jdconrad replied:
Updated the tests to the new namespace.

martijnvg commented on core/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestDeleteStoredScriptAction.java:
```diff
@@ -35,24 +36,33 @@
-        return request.param("lang");
+        // Note {lang} is actually {id} in the first handler.  It appears
+        // parameters as part of the path must be of the same ordering relative
+        // to name or they will not work as expected.
```
true... I also don't see a way to make this nicer.

jdconrad replied:
:(, but it will be removed in master anyway

martijnvg commented on core/src/main/java/org/elasticsearch/script/StoredScriptSource.java:
```diff
@@ -0,0 +1,511 @@
+ * {@link StoredScriptSource} represents user-defined parameters for a script
+ * saved in the {@link ClusterState}.
+ */
+public class StoredScriptSource extends AbstractDiffable<StoredScriptSource> implements Writeable, ToXContent {
```
:+1: much better than script as bytes

jdconrad replied:
:)

martijnvg commented on core/src/test/java/org/elasticsearch/script/StoredScriptTests.java:
```diff
@@ -0,0 +1,333 @@
+        assertThat(rnfe.getMessage(), equalTo("stored script [test] does not exist and cannot be deleted"));
+    }
+
+    public void testSourceParsing() throws Exception {
```
@colings86 is adding base test classes for writeable and xcontent serialization / de-serialization via #22281

That should get merged quickly. When that is in can you make this test class extend from `AbstractSerializingTestCase`. That will add random serialization tests in addition that the serialization  that exist already.

martijnvg replied:
and maybe do the same for `ScriptMetaDataTests`?

jdconrad replied:
Done.

martijnvg commented on core/src/main/java/org/elasticsearch/script/ScriptMetaData.java:
```diff
@@ -18,168 +18,414 @@
+
+                    int split = id.indexOf('#');
+
+                    if (split == -1) {
```
I think we can move the split check to the place where we assign the `id` variable?

jdconrad replied:
split is used in both the if and else blocks

jdconrad commented:
@martijnvg Thank you so much for reviewing this!  I agree that the old API should be removed in master -- the plan was to get this in, backport it to version 5.2 (although maybe 5.3 is more realistic given the upcoming holidays), and then remove the old API in master where we can.
I'll get started working on your other comments pretty soon.

jdconrad commented on core/src/main/java/org/elasticsearch/script/ScriptMetaData.java:
```diff
@@ -18,168 +18,414 @@
+         */
+        public Builder storeScript(String id, StoredScriptSource source) {
+            StoredScriptSource previous = scripts.put(id, source);
+            scripts.put(source.getLang() + "#" + id, source);
```
I thought about this as well and it certainly makes sense, but the tricky part is the user has no choice for stored templates.  I wanted to make sure that the user could still have a stored template with the same name as a script for this version so they have time to come up with an appropriate plan for the deprecation.  With the correct planning the old scheme will simply disappear when loading the scripts from cluster state metadata because we can make it ignore (and print a warning) that an old script cannot be loaded similar to how file scripts work now for scripts that do not compile.

Also, right now this technically isn't a breaking change which I think supports the user a little better (though, hopefully no one is dependent on having scripts of the same name with different languages).

jdconrad commented on core/src/main/java/org/elasticsearch/script/ScriptMetaData.java:
```diff
@@ -18,168 +18,414 @@
+        // Version 5.2+ will output the contents of the scripts' Map using
+        // StoredScriptSource to stored the language, code, and options.
+        if (out.getVersion().onOrAfter(Version.V_5_2_0_UNRELEASED)) {
+            out.write(scripts.size());
```
Nice catch!  :)

jdconrad commented on docs/reference/migration/migrate_6_0/scripting.asciidoc:
```diff
@@ -1,6 +1,8 @@
 === Scripting changes
 
+
+
```
Accidental.  Good catch!

jdconrad commented:
Added tests for rolling upgrade.  These will need to be updated when the code is backported to 5.3 and when 6.0 deletes the code for the old API.

jdconrad commented:
@martijnvg I think this is ready for another round of review.  I believe I addressed all the comments, and added several tests that you requested.  Thanks!

jdconrad commented:
@martijnvg I added the patch you sent me to address the bwc bugs you found with one minor change --  I decided to clean up the Script readFrom and writeTo a little bit to separate out each of the read and write versions into their own if/else block as I think that will help clear up potential future confusion.  And thanks again so much for catching some of the bwc bugs!  Ready for another round of review when you have the chance.

jdconrad commented:
@martijnvg Thanks for all the reviews here!  Added this as a follow up issue (#22887)  I will commit this code once I get master integrated properly and all the tests passing.

clintongormley commented:
@jdconrad in 6.0 i assume the namespace support will no longer be deprecated but will be removed?  If so, please add a note to the migration docs for 6.0

jdconrad commented:
@clintongormley Yes, it will be removed, but that requires another change to actually remove it from 6.0 that I'll try to follow up with this week.  Right now, it's still technically deprecated there, but not removed.  Once, I make the change I'll make sure to add that as a (breaking?) note to the migration docs.

Mpdreamz commented:
@jdconrad should the spec for script api's in `master` be changed from:

```json
      "path": "/_scripts/{lang}",
      "paths": [ "/_scripts/{lang}", "/_scripts/{lang}/{id}" ],
```

to

```json
      "path": "/_scripts/{id}",
      "paths": [ "/_scripts/{id}" ],
```

Or has the removal not made it in yet to `master` ?
